Libraries/Opc.Ua.PubSub/Encoding/JsonNetworkMessage.cs

```diff
@@ -0,0 +1,471 @@
+                        jsonDataSetMessage.SetMessageContentMask((JsonDataSetMessageContentMask)jsonMessageSettings.DataSetMessageContentMask);
+                        jsonDataSetMessage.SetFieldContentMask((DataSetFieldContentMask)dataSetReader.DataSetFieldContentMask);
+                        // set the flag that indicates if dataset message shall have a header
+                        jsonDataSetMessage.HasDataSetMessageHeader = (networkMessageContentMask & JsonNetworkMessageContentMask.DataSetMessageHeader) != 0;
+
+                        jsonDataSetMessage.DecodePossibleDataSetReader(jsonDecoder, messagesList.Count, messagesListName, dataSetReader);
+                        if (jsonDataSetMessage.DataSet != null)
+                        {
+                            m_uaDataSetMessages.Add(jsonDataSetMessage);
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Unexpected exception in DecodeSubscribedDataSets
+                Utils.Trace(ex, "JsonNetworkMessage.DecodeSubscribedDataSets");
+            }
+        }
+        #endregion
+
+        #region Private Methods - Encoding
+        /// <summary>
+        /// Encodes the object in a binary stream.
+        /// </summary>
+        private void Encode(JsonEncoder jsonEncoder)
+        {
+            if (jsonEncoder == null)
+            {
+                throw new ArgumentException(nameof(jsonEncoder));
+            }
+
+            if (HasNetworkMessageHeader)
+            {
+                EncodeNetworkMessageHeader(jsonEncoder);
+            }
+            EncodeMessages(jsonEncoder);
+            EncodeReplyTo(jsonEncoder);
+        }
+
+        /// <summary>
+        ///  Encode Network Message Header
+        /// </summary>
+        private void EncodeNetworkMessageHeader(JsonEncoder jsonEncoder)
+        {
+            jsonEncoder.WriteString(FieldMessageId, MessageId);
+            jsonEncoder.WriteString(FieldMessageType, MessageType);
+
+            if ((NetworkMessageContentMask & JsonNetworkMessageContentMask.PublisherId) != 0)
+            {
+                jsonEncoder.WriteString(FieldPublisherId, PublisherId);
+            }
+
+            if ((NetworkMessageContentMask & JsonNetworkMessageContentMask.DataSetClassId) != 0)
+            {
+                jsonEncoder.WriteString(FieldDataSetClassId, DataSetClassId);
+            }
+        }
+
+        /// <summary>
+        /// Encode DataSetMessages
+        /// </summary>
+        private void EncodeMessages(JsonEncoder encoder)
+        {
+            if (DataSetMessages != null && DataSetMessages.Count > 0)
+            {
+                if (HasSingleDataSetMessage)
+                {
+                    // encode single dataset message                   
+                    JsonDataSetMessage jsonDataSetMessage = DataSetMessages[0] as JsonDataSetMessage;
+                    if (jsonDataSetMessage != null)
```

Comment:
how is the case with and without datasetmessageheader handled here?

Reply:
`EncodeMessages()` is not called in case datasetmessageheader  is missing.
It encodes only the payload here:

```
if (!jsonDataSetMessage.HasDataSetMessageHeader)
  {
      // (UA Specs:) If the NetworkMessageHeader and the DataSetMessageHeader bits are not set
      // and SingleDataSetMessage bit is set, the NetworkMessage
      // is a JSON object containing the set of name/value pairs defined for a single DataSet.
      jsonDataSetMessage.EncodePayload(encoder, false);
  }
```